src/op_mode/file.py

```diff
@@ -0,0 +1,378 @@
+                else:
+                    print_error(f'File or directory {path} not found.')
+                    sys.exit(1)
+                sys.stdout.flush()
+                # Call `less(1)` and wait for it to terminate before going forward.
+                cmd(['/usr/bin/less', '-X', f.name], stdout=sys.stderr)
+        # The stream to the temporary file could break for any reason.
+        # It's much less fragile than if we streamed directly to the process stdin.
+        # But anything could still happen and we don't want to scare the user.
+        except (BrokenPipeError, EOFError, KeyboardInterrupt, OSError):
+            fix_terminal()
+            sys.exit(1)
+
+def show(type: str, path: str) -> None:
+    if type == 'remote':
+        with tempfile.NamedTemporaryFile(delete=False) as temp:
+            download(temp.name, path)
+            show_locally(temp.name)
+    elif type == 'image':
+        show_locally(parse_image_path(path))
+    elif type == 'local':
+        show_locally(path)
+    else:
+        print_error(f'Unknown target for showing: {type}')
+        print_error('Valid types are "remote", "image" and "local".')
+        sys.exit(1)
+
+
+## Copying procedures
+def copy(source_type: str, source_path: str,
+         destination_type: str, destination_path: str) -> None:
+    """
+    Copy a file or directory locally, remotely or to and from an image.
+    Directory uploads and downloads not supported.
+    """
+    source = ''
+    try:
+        # Download to a temporary file and use that as the source.
+        if source_type == 'remote':
+            source = tempfile.NamedTemporaryFile(delete=False).name
+            download(source, source_path)
+        # Prepend the image root to the path.
+        elif source_type == 'image':
+            source = parse_image_path(source_path)
+        elif source_type == 'local':
+            source = source_path
+        else:
+            print_error(f'Unknown source type: {source_type}')
+            print_error(f'Valid source types are "remote", "image" and "local".')
+            sys.exit(1)
+
+        # Directly upload the file.
+        if destination_type == 'remote':
+            if os.path.isdir(source):
+                print_error(f'Cannot upload {source}. Directory uploads not supported.')
+                sys.exit(1)
+            upload(source, destination_path)
+        # No need to duplicate local copy operations for image copying.
+        elif destination_type == 'image':
+            copy('local', source, 'local', parse_image_path(destination_path))
+        # Try to preserve metadata when copying.
+        elif destination_type == 'local':
+            if os.path.isdir(source):
+                if os.path.isdir(destination_path):
+                    destination_path = os.path.join(destination_path, os.path.basename(source))
+                shutil.copytree(source, destination_path, copy_function=zealous_copy)
+            else:
```

Comment:
A missing check here causes problems, in applying 'chown' to the destination directory:
```
diff --git a/src/op_mode/file.py b/src/op_mode/file.py
index b09631ba2..b675afebf 100755
--- a/src/op_mode/file.py
+++ b/src/op_mode/file.py
@@ -244,6 +244,8 @@ def copy(source_type: str, source_path: str,
                     destination_path = os.path.join(destination_path, os.path.basename(source))
                 shutil.copytree(source, destination_path, copy_function=zealous_copy)
             else:
+                if os.path.isdir(destination_path):
+                    destination_path = os.path.join(destination_path, os.path.basename(source))
                 zealous_copy(source, destination_path)
         else:
             print_error(f'Unknown destination type: {source_type}')
```

Reply:
I moved the shared check outside the conditional for simplicity.